### PR TITLE
45 configuring spring security jwt add role

### PR DESCRIPTION
**Pull request by SMSaeedi**

### CODE REVIEW

File: src/main/kotlin/com/linchpino/core/enums/AccountTypeEnum.kt [hunk 1]
```diff
@@ -1,7 +1,8 @@
 package com.linchpino.core.enums
 
 enum class AccountTypeEnum(val value: Int) {
```
Comment (m-tilab) [inline src/main/kotlin/com/linchpino/core/enums/AccountTypeEnum.kt:3]: Please revert this change and enable the test if possible

File: src/main/kotlin/com/linchpino/core/repository/AccountRepository.kt [hunk 2]
```diff
@@ -29,7 +29,7 @@ interface AccountRepository : JpaRepository<Account, Long>{
     JOIN
```
Comment (m-tilab) [inline src/main/kotlin/com/linchpino/core/repository/AccountRepository.kt]: This change can be done in another task. Please revert it!
Related Task:
https://trello.com/c/CSRb35jF/72-update-closestmentortimeslots-native-query-so-that-all-constant-values-be-passed-as-argument-instead-of-hardcode-it

File: src/main/kotlin/com/linchpino/core/enums/AccountTypeEnum.kt (hunk 1 above)
Comment (SMSaeedi) [reply src/main/kotlin/com/linchpino/core/enums/AccountTypeEnum.kt:3]: why???

**Review by m-tilab — COMMENTED** [review]
Dear @sharifyy Can you review this PR

File: src/test/kotlin/com/linchpino/core/controller/AccountControllerTestIT.kt [hunk 3]
```diff
@@ -78,7 +78,7 @@ class AccountControllerTestIT {
 
     @Test
     fun `test creating account with blank lastName results in bad request`() {
-        val invalidRequest = CreateAccountRequest("John", "", "john.doe@example.com", "secret", 1)
+        val invalidRequest = CreateAccountRequest("John", "", "john.doe@example.com", "secret", 2)
```
Comment (m-tilab) [inline src/test/kotlin/com/linchpino/core/controller/AccountControllerTestIT.kt]: Maybe it would be better to use "AccountTypeEnum.MENTOR.value" instead of number 2

File: src/main/kotlin/com/linchpino/core/security/SecurityUser.kt [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
-        return mutableListOf<SimpleGrantedAuthority>()
+class SecurityUser(
+    private val roles: Set<Role>,
+    private val email: String,
+    private val password: String,
+    private val status: AccountStatusEnum
+) : UserDetails {
+    override fun getAuthorities(): List<GrantedAuthority> {
+        return roles.map {
+            SimpleGrantedAuthority("SCOPE_${it.roleName.value}")
```
Comment (m-tilab) [inline src/main/kotlin/com/linchpino/core/security/SecurityUser.kt]: Can you tell me where we defined the "it.roleName.value" value?

File: src/main/kotlin/com/linchpino/core/security/UserService.kt [hunk 5]
```diff
@@ -1,38 +1,21 @@
...
 
 @Service
+@Transactional
 class UserService(private val accountRepository: AccountRepository) : UserDetailsService {
 
+    @Transactional(readOnly = true)
     override fun loadUserByUsername(username: String): UserDetails =
-        accountRepository.findByEmailIgnoreCase(username)?.let { SecurityUser(it) }
+        accountRepository.findByEmailIgnoreCase(username)
+            ?.let { SecurityUser(it.roles(), it.email, it.password, it.status) }
```
Comment (m-tilab) [inline src/main/kotlin/com/linchpino/core/security/UserService.kt:17]: what is the meaning of "?.let" ?

File: src/main/kotlin/com/linchpino/core/security/SecurityUser.kt (hunk 4 above)
Comment (SMSaeedi) [reply src/main/kotlin/com/linchpino/core/security/SecurityUser.kt]: AccountTypeEnum which is the type of roleName. To access the value property of roleName when it is defined, in  Account entity

File: src/test/kotlin/com/linchpino/core/controller/AccountControllerTestIT.kt (hunk 3 above)
Comment (SMSaeedi) [reply src/test/kotlin/com/linchpino/core/controller/AccountControllerTestIT.kt]: To avoid modifications on test class, using AccountTypeEnum.MENTOR.code will do.

File: src/main/kotlin/com/linchpino/core/security/UserService.kt (hunk 5 above)
Comment (SMSaeedi) [reply src/main/kotlin/com/linchpino/core/security/UserService.kt:17]: If nullableValue is not null, the block of code inside let { ... } is executed, and the non-null value of nullableValue is passed as it.

**Review by m-tilab — COMMENTED** [review]
image
This test has failed for me. Can you look at it?

**Review by m-tilab — COMMENTED** [review]
Did you add some new tests to assess your changes?

**Review by m-tilab — APPROVED** [review]
